File: neuron-di-guice/src/test/java/global/tranquillity/neuron/di/guice/it/FooBarModuleTestSuite.java
package global.tranquillity.neuron.di.guice.it;

import com.google.inject.Module;
import global.tranquillity.neuron.di.api.Neuron;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

@Neuron
public interface FooBarModuleTestSuite extends ModuleTestSuite {

    @Override
    default Module module() { return new FooBarModule(); }

    default void test() {
        final Bar bar1 = getInstance(Bar.class);
        final Bar bar2 = getInstance(Bar.class);
        assertThat(bar1, is(not(sameInstance(bar2))));
        assertThat(bar1, is(instanceOf(BarImpl.class)));
        assertThat(bar2, is(instanceOf(BarImpl.class)));
        assertThat(bar1.foo(), is(instanceOf(FooImpl.class)));
        assertThat(bar1.foo(), is(sameInstance(bar2.foo())));
        assertThat(bar1.foo().i(), is(1));
    }
}

File: neuron-di-guice/src/test/scala/global/tranquillity/neuron/di/guice/scala/it/FooBarModuleTestSuite.scala
package global.tranquillity.neuron.di.guice.scala.it

import javax.inject.Singleton

import global.tranquillity.neuron.di.api.Neuron
import global.tranquillity.neuron.di.guice.it.{Bar, BarImpl, Foo, FooImpl, FooBarModuleTestSuite => jFooBarModuleTestSuite}
import global.tranquillity.neuron.di.guice.scala.{NeuronModule, _}

@Neuron
trait FooBarModuleTestSuite extends jFooBarModuleTestSuite {

  def module = new NeuronModule {

    def configure() {
      bindConstantNamed("one").to(1)
      bindClass[Foo]
        .named("impl")
        .toClass[FooImpl]
        .inScope[Singleton]
      bindClass[Bar].toClass[BarImpl]
    }
  }
}
